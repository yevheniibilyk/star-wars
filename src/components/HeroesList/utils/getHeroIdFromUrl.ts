export default function getHeroIdFromUrl(url: string) {
  const idString = url.replace(/.+\/(\d+)\/$/, '$1');

  if (!idString) {
    return url;
  }

  return Number.parseInt(idString, 10);
}