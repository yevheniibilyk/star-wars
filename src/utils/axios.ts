import axios from 'axios';
import { apiUrl } from '../config';

const axiosInstance = axios.create({
  baseURL: apiUrl
});

export default axiosInstance;