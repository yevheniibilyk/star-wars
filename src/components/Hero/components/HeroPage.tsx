import React, { Component } from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';
import { CircularProgress } from '@material-ui/core';
import { Hero } from '../../../types/Hero';
import { axios } from '../../../utils';
import HeroCard from './HeroCard';
import '../styles/_hero-page.scss';

type HeroPageState = {
  hero?: Hero;
  loading: boolean,
  error: boolean
};

class HeroPage extends Component<RouteComponentProps, HeroPageState> {
  constructor(props: RouteComponentProps) {
    super(props);

    this.state = {
      hero: void 0,
      loading: false,
      error: false
    };
  }

  componentDidMount() {
    const { match: { params } } = this.props;

    //@ts-ignore
    this.getHeroById(params.id);
  }

  getHeroById = async (id: string) => {
    try {
      this.setState({ loading: true });

      const { data: hero }: any = await axios.get(`/hero/${id}`);

      this.setState({ hero });
    } catch (e) {
      this.setState({ error: true })
    } finally {
      this.setState({ loading: false });
    }
  }

  render() {
    const { loading, hero, error } = this.state;

    if (loading) {
      return <CircularProgress className="hero-page__loader" />;
    }

    if (error || !hero) {
      return (
        <div className="hero-page__not-found">
          Not Found
        </div>
      );
    }

    return (
      <div className="hero-page">
        <HeroCard hero={hero} />
      </div>
    );
  }
}

export default withRouter(HeroPage);
