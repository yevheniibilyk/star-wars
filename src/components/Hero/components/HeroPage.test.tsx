import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import HeroPage from './HeroPage';
import { axios } from '../../../utils';

const MOCK_HERO = {
  name: 'Luke Skywalker',
  height: '172',
  gender: 'male',
  mass: '83',
  skin_color: 'white',
  eye_color: 'blue',
  birth_year: '1999'
}

jest.mock('../../../utils/axios', () => ({
  get: jest.fn(() => Promise.resolve({ data: MOCK_HERO }))
}))

function renderHeroPage(props = {}) {
  return render(
    <BrowserRouter>
      <HeroPage {...props} />
    </BrowserRouter>
  )
}

describe('HeroPage', () => {
  test('should be able to render HeroPage', async () => {
    // @ts-ignore
    axios.get.mockImplementationOnce(() => Promise.resolve({ data: MOCK_HERO }));

    const { container } = await renderHeroPage({ match: { params: { id: 2 } } });

    expect(container).toMatchSnapshot();
    expect(screen.getByText(MOCK_HERO.name)).toBeInTheDocument();
    expect(axios.get).toHaveBeenCalled();
  });

  test('should be able to handle request errors', async () => {
    // @ts-ignore
    axios.get.mockImplementationOnce(() => Promise.reject(new Error('404')));

    await renderHeroPage({ match: { params: { id: 2 } } });

    expect(screen.getByText('Not Found')).toBeInTheDocument();
    expect(axios.get).toHaveBeenCalled();
  });
});
