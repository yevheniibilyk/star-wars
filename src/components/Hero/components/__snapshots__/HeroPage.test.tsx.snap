// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroPage should be able to render HeroPage 1`] = `
<div>
  <div
    class="hero-page"
  >
    <div
      class="MuiPaper-root MuiCard-root hero-card MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardMedia-root hero-card__image"
        style="background-image: url(s-l300.jpg);"
      />
      <div
        class="MuiCardContent-root hero-card__content"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4"
        >
          Luke Skywalker
        </h4>
        <div
          class="hero-card__param"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Height
            :
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            172
          </p>
        </div>
        <div
          class="hero-card__param"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Gender
            :
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            male
          </p>
        </div>
        <div
          class="hero-card__param"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Mass
            :
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            83
          </p>
        </div>
        <div
          class="hero-card__param"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Skin color
            :
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            white
          </p>
        </div>
        <div
          class="hero-card__param"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Eye color
            :
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            blue
          </p>
        </div>
        <div
          class="hero-card__param"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            Birth year
            :
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
          >
            1999
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
